### ISSUE TITLE
imgui.image does not work if texture_id exceeds 256

### ISSUE THREAD

**Issue by mCodingLLC** [pyimgui/pyimgui#248]
I'm using pyimgui with glfw on Wayland using EGL on an Ubuntu VM. For some reason this code crashes when:

1. at least one image has been drawn before, AND
2. the `texture_id` exceeds 256.

The crash happens because somehow a `texture_id` of a `DrawCmd` gets set to another `DrawCmd` object instead of an int. It doesn't matter how many textures are generated (300 in the example below) or what you start `texture_id` at, it always crashes when `texture_id` is 257, or if the first loop iteration is 257 or higher then it will crash on the second iteration.  The error happening at 256+1 is obviously suspicious.

Here is the traceback:
```python3
Traceback (most recent call last):
  File "/home/mcoding/PycharmProjects/pyimgui-demo/pyimgui-demo.py", line 77, in <module>
    main()
  File "/home/mcoding/PycharmProjects/pyimgui-demo/pyimgui-demo.py", line 41, in main
    impl.render(imgui.get_draw_data())
  File "/home/mcoding/PycharmProjects/pyimgui-demo/venv/lib/python3.9/site-packages/imgui/integrations/opengl.py", line 232, in render
    gl.glBindTexture(gl.GL_TEXTURE_2D, command.texture_id)
ctypes.ArgumentError: argument 2: <class 'TypeError'>: wrong type

```

And here is a minimal source causing the traceback:

```python3
# -*- coding: utf-8 -*-
import glfw
import OpenGL.GL as gl
import imgui
from imgui.integrations.glfw import GlfwRenderer


def main():
	imgui.create_context()
	window = impl_glfw_init()

	impl = GlfwRenderer(window)

	ids = gl.glGenTextures(300)
	print(ids)

	texture_id = 0
	while not glfw.window_should_close(window):
		glfw.poll_events()
		impl.process_inputs()
		imgui.new_frame()

		if imgui.begin_main_menu_bar():
			if imgui.begin_menu("File", True):
				clicked_quit, selected_quit = imgui.menu_item("Quit", 'Cmd+Q', False, True)

				if clicked_quit:
					exit(1)

				imgui.end_menu()
			imgui.end_main_menu_bar()

		print(texture_id)
		imgui.image(texture_id, 640, 480)
		texture_id = min(texture_id + 1, ids[-1])

		gl.glClearColor(1., 1., 1., 1)
		gl.glClear(gl.GL_COLOR_BUFFER_BIT)

		imgui.render()
		impl.render(imgui.get_draw_data())
		glfw.swap_buffers(window)

	impl.shutdown()
	glfw.terminate()


def impl_glfw_init():
	width, height = 1280, 720
	window_name = "minimal ImGui/GLFW3 example"

	if not glfw.init():
		print("Could not initialize OpenGL context")
		exit(1)

	# OS X supports only forward-compatible core profiles from 3.2
	glfw.window_hint(glfw.CONTEXT_VERSION_MAJOR, 3)
	glfw.window_hint(glfw.CONTEXT_VERSION_MINOR, 3)
	glfw.window_hint(glfw.OPENGL_PROFILE, glfw.OPENGL_CORE_PROFILE)
	glfw.window_hint(glfw.OPENGL_FORWARD_COMPAT, gl.GL_TRUE)

	# Create a windowed mode window and its OpenGL context
	window = glfw.create_window(
		int(width), int(height), window_name, None, None
	)
	glfw.make_context_current(window)

	if not window:
		glfw.terminate()
		print("Could not initialize Window")
		exit(1)

	return window


if __name__ == "__main__":
	main()

```

I run the command with

```sh
PYOPENGL_PLATFORM=egl python pyimgui-demo.py 
```

but I'm not sure if this is an EGL thing or not.

**Comment by learn-more** [comment]
Thanks for your reproduction case, it reproduces perfectly!

Here is my analysis:
* Add code to `ImGui::Image` printing the `user_texture_id`:
  `printf("%p\n", user_texture_id);`
  output:
  ```
  000001B3A67E6910
  000001B3A67E6930
  000001B3A67E6950
  000001B3A67E6970
  and so on
  ```
* Observe that none of the arguments here look like an int, so they are most likely python objects representing an int
  This actually makes sense, considering everything is an object in python.
  So what is probably happening here, is that we give a python object to ImGui, but it does not know that it should use reference counting for that object, since all ImGui knows is that it gets a `void*` pointer.

* Now why does the higher number not work, but lower numbers do?
  The last lines of the output give a hint:
  ```
  000001B3A6816950
  000001B3A6816970
  000001B3A6816990
  000001B3BF39BC50
  ```
  Notice the jump in addresses for the last item?
  **My educated guess here is that python pre-allocates all objects for integers below `257`, and that these objects are never involved in refcounting, for optimization purposes.**

In short: We need to tell cpython to properly add a reference to data that we pass to the `user_texture_id` argument, and we should find another place to dereference it.

**Comment by mCodingLLC** [comment]
I'm not familiar enough with the imgui implementation to know how to fix this, but you seem spot on with the Python stuff. Indeed, the integers -5 to 256 in CPython are pre-allocated and treated as singletons. I'm not sure whether small integers participate in refcounting, but it seems reasonable that they don't, or at the very least that they are never deallocated during the lifetime of the interpreter. gc.is_tracked appears to return False for all int objects, even large ones.

**Comment by mCodingLLC** [comment]
(Just guessing) another possible solution would be to convert what you get for user_texture_id to a C int as it comes in, and back to a python int whenever it is accessed from Python, which Cython has a mechanism for doing. Just mark the variable as unsigned int instead of void* and I believe Cython does this automatically.

**Comment by mCodingLLC** [comment]
Tried this on my own machine and it seems an easy fix is to convert the Python ints to and from uintptr_t (int type with same size as void pointer) at the Python interface boundary. E.g.

```python3
cdef class _FontAtlas(object):
    # ... snipped
    @property
    def texture_id(self):
        return <uintptr_t>self._ptr.TexID

    @texture_id.setter
    def texture_id(self, uintptr_t value):
        self._ptr.TexID = <void *> value
```
Making these changes throughout the core.pyx file appears to fix everything on my end at least on the application I'm working on. However, I am assuming here that these texture_ids are always going to be Python ints and not some other kind of object. Is that a reasonable thing to assume? I can submit a PR if that approach sounds good to you.

**Comment by mCodingLLC** [comment]
I've also confirmed that adding Py_INCREF fixes the crash as well. However, this creates a memory leak (well, I suppose technically it just creates an int object that will never be garbage collected) as there is no corresponding Py_DECREF. Since the cimgui.Image call is C code I don't see where the corresponding Py_DECREF could go.
```python3
from cpython.ref cimport Py_INCREF

# ... snip

def image(
    texture_id,
    float width,
    float height,
    tuple uv0=(0, 0),
    tuple uv1=(1, 1),
    tuple tint_color=(1, 1, 1, 1),
    tuple border_color=(0, 0, 0, 0),
):
    Py_INCREF(texture_id)
    cimgui.Image(
        <void*>texture_id,
        _cast_args_ImVec2(width, height),  # todo: consider inlining
        _cast_tuple_ImVec2(uv0),
        _cast_tuple_ImVec2(uv1),
        _cast_tuple_ImVec4(tint_color),
        _cast_tuple_ImVec4(border_color),
    )
```

**Comment by learn-more** [comment]
@KinoxKlark What do you suggest?

**Comment by KinoxKlark** [comment]
> Tried this on my own machine and it seems an easy fix is to convert the Python ints to and from uintptr_t (int type with same size as void pointer) at the Python interface boundary. 

This feels to me like the correct thing to do indeed. You are welcome to submit a PR for this fix. However, be aware of the other places where this "pattern" and the "texture_id" are passed and where this bug may arise again. By a quick search through `core.pyx` (in dev/version-2.0 branch) I found
* `_DrawList.push_texture_id()`
* `_DrawList.add_image()`
* `image_button()`
* `image()`

And I suppose a similar error hides in `_DrawCmd.texture_id`. It would be nice if everything was consistent with respect to that fix. 

I may have overlooked some other places where this should be fixed, but propose a PR with those and we'll check everything during reviewing.

Thanks!

**Comment by KinoxKlark** [comment]
This has been fixed by #250.
